src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -317,7 +310,12 @@ private void setCachedResultAndResumeToken() {
         }
 
         try {
-            return Optional.ofNullable(cursor.tryNext());
+            BsonDocument next = cursor.tryNext();
+            if (next == null && cursor.getServerCursor() == null) {
```

Comment:
Never thought to use `getServerCursor()` like this.  Would `isClosed` work as well?  I think the contract is that after a tryNext/next where the server doesn't return a cursorId, the MongoCursor closes itself.

Reply:
The issue is the cursor isn't closed, if it was it would have thrown an illegal state exception. So that could be a bug in the java driver.

Also if we fix that then we would need to add an `isClosed` method on the change stream cursor or the `MongoCursor`.

Reply:
Right, I didn't realize it didn't exist.